lib/internal/modules/esm/loader.js

```diff
@@ -69,19 +82,24 @@ class Loader {
     if (!isMain)
       validateString(parentURL, 'parentURL');
 
-    const resolved = await this._resolve(specifier, parentURL, defaultResolve);
+    const resolved = await this._resolve(
+      { specifier, parentURL }, defaultResolve, this);
 
     if (typeof resolved !== 'object')
       throw new ERR_INVALID_RETURN_VALUE(
         'object', 'loader resolve', resolved
       );
 
     const { url, format } = resolved;
-
     if (typeof url !== 'string')
       throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
         'string', 'loader resolve', 'url', url
       );
+    if (!format && new URL(url).protocol === 'file:') {
+      throw new ERR_UNKNOWN_FILE_EXTENSION(
+        extname(new URL(url).pathname),
+        fileURLToPath(url));
+    }
```

Comment:
Can we please move this exact error message into the `getFormat` function.

If we want the error here, it should be a far more generic hook error like `The getFormat hook did not return a valid module format.`